File: graph/27.cpp
#include <bits/stdc++.h>

using namespace std;

using Graph = vector<vector<int>>;

bool topologicalSort(const Graph &G, vector<int> inDegree, vector<int> &order) {
    queue<int> q;

    for(int i = 0; i < inDegree.size(); ++i) {
        if(inDegree[i] == 0)
            q.push(i);
    }

    while(!q.empty()) {
        auto v = q.front();
        q.pop();

        order.push_back(v);

        for(const auto& u: G[v]) {
            inDegree[u]--;
            if(inDegree[u] == 0)
                q.push(u);
        }
    }


    return accumulate(inDegree.begin(), inDegree.end(), 0) == 0;
}

void createListAdj(Graph &G, int source, int dest) {
    G[source].push_back(dest);
}

vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
    Graph G(numCourses);
    vector<int> answer;
    vector<int> inDegree(numCourses, 0);

    for(const auto& item: prerequisites) {
        createListAdj(G, item[1], item[0]);
        inDegree[item[0]]++;
    }

    if(topologicalSort(G, inDegree, answer))
        return answer;

    return {};
}

int main() {
    int numCourses = 2;
    vector<vector<int>> p = {{1, 0}};

    for(const auto& e: findOrder(numCourses, p))
        cout<<e<<" ";


    return 0;
}

File: graph/27.py
from typing import List
from collections import deque


def topological_sort(graph: List[List], in_degree: List, order: List):
    q = deque()
    for index, value in enumerate(in_degree):
        if value == 0:
            q.append(index)

    while len(q) > 0:
        v = q.popleft()
        order.append(v)

        for u in graph[v]:
            in_degree[u] -= 1

            if in_degree[u] == 0:
                q.append(u)

    return sum(in_degree) == 0


def create_adj_matrix(graph: List[List], source: int, dest: int):
    graph[source].append(dest)


def findOrder(numCourses: int, prerequisites: List[List[int]]) -> List[int]:
    graph = [[] for _ in range(numCourses)]
    in_degree = [0 for _ in range(numCourses)]
    order = []

    for item in prerequisites:
        create_adj_matrix(graph, item[1], item[0])
        in_degree[item[0]] += 1

    if topological_sort(graph, in_degree, order) is True:
        return order

    return []


numCourses = 2
prerequisites = [[1, 0]]
print(findOrder(numCourses, prerequisites))
